### PR TITLE
Improve shortcut key reporting

### LINKED ISSUE #10807: Outlook 2010: No Shortcut Key Given for the To Button in the New Message Window
### Steps to reproduce:

From within Outlook 2010, open a new message and shift + tab to the "To..." button

### Actual behavior:

When NVDA encounters the button, it outputs the following: "To... button alt +" - it is missing part of the shortcut key

### Expected behavior:

NVDA should either output the entirety of the shortcut key for this button (if one actually exists), or simply ignore the "alt +" suffix for this particular button

### PR DESCRIPTION
**Pull request by CyrilleB79**
### Link to issue number:
Fixes #10807

### Summary of the issue:
When reported by speech, some shortcut key or combination cannot be understood correctly in the following cases:
* A punctuation may not be reported due to symbol level being too low, e.g. "alt+." reported as "alt plus" instead of "alt plus dot" in Outlook.
* A character reported as a word instead of being reported as a character. E.g. in French, the letter "à" (a grave) is a word; in MS Word, the shortcut "alt+à" is reported as "alt plus a" instead of "alt plus a grave". Same for "Y" which is a word in French and the name of the letter is "i grec" whereas the word just reads "i" ("ee").

### Description of user facing changes
Shortcut key reporting will be improved in some cases. This applies to on-demand reporting (shift+numpad2) as well as when "Report Object Shortcut Keys" option is checked in Object presentation settings.

The cases where it should be improved are:
* punctuation key, e.g. "." (dot)
* letter keys when the name of the letter is not spoken the same way as the equivalent one-letter word, e.g. "à" or "y" in French.

Braille reporting should remain unchanged.

### Description of development approach
* Generate a speech sequence to report shortcut keys from the shortcut key string provided by the API. This shortcut key string may contain one or two shortcuts. When there are two shortcuts, they are separated by two spaces.
* Each shortcut is split into its component keys.
* If a key is a punctuation, symbol reporting processing is used so that the punctuation can be reported no matter the punctuation level.
* If a key is a single character, `CharacterModeCommand` is passed to the synth so that the name of the letter is reported and not a possible one-letter word.

Note: all the processing to split shortcut keys is in `speech/speech.py` as private functions. Reviewer: should they remain here or would they fit better elsewhere?

### Testing strategy:
Manual tests for the following shortcuts:
* Sequential shortcut mixing comma and space separators: Word (French): "Alt, F, L, Y 2"
* Two shortcuts for the same control, one sequential and another simultaneous: Word (French): "Alt, É, É  Ctrl+F"
* Two shortcuts for the same control, one sequential and another simultaneous using the "+" (plus) key: Word (French): "Alt, L, 6  Ctrl+Maj++"
* Letter with diacritic alone: NVDA menu -> Help -> About (French): "à" from "À propos..."
* Letter with diacritic combined with modifier: Outlook - Message writing: "Alt+à"
* Shortcut with punctuation alone: "?" used as help menu in 6pad++.
* Shortcut with punctuation combined with modifier: Outlook's message (writing): "Alt+."

Tested the shortcut "Touche de logo Windows + Plus (+)" which is from Windows Magnifier. This shortcut has its shortcut keys separated with " + " (with space around the "plus" symbol) and uses the "+" symbol in a key label. It was tested as follows in the console:
```
>>> from speech.shortcutKeys import getKeyboardShortcutsSpeech
>>> getKeyboardShortcutsSpeech('Touche de logo Windows + Plus (+)')
['Touche de logo Windows + Plus (+)']
```
Direct test on the magnifier cannot be done running from source since the Magnifier UI is not accessible when running from source (maybe UI Access is required for it?)

Tested also with various voices (Eloquence, OneCore, SAPI5 (Windows), eSpeak.

System tests have been updated to pass since the number of spaces in shortcut keys has changed. It is due to the use of `CharacterModeCommand` in the speech sequence which adds some extra spaces when this sequence is converted to string.

### Known issues with pull request:
* Seen with OneCore that "é" in French is not reported as "e accent aigu" ("e acute") although the `CharacterModeCommand` is used.
This PR does not fix this which is more a OneCore issue.
* Shortcut keys as reported in the speech viewer may have some extra spaces with respect to what was previously reported. It is due to the usage of `CharacterModeCommand` in the speech sequence which converts to extra space when this sequence is converted to string for the speech viewer.

### Change log entries:
Bug fixes
`Reporting of object shortcut keys has been improved. (#10807)`
### Code Review Checklist:

- [x] Pull Request description:
  - description is up to date
  - change log entries
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] API is compatible with existing add-ons.
- [x] Documentation:
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] UX of all users considered:
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
- [x] Security precautions taken.

### CODE REVIEW

**Comment by cary-rowen** [comment]
Hi @CyrilleB79 
thank you for your efforts.

Can the fix of this PR cover the situation I described below:
If I check "Report object shortcut keys", NVDA will automatically guess the shortcuts for some menu items in the Chinese interface.
For example, on the "Recycle Bin" icon on the desktop, there is an item called "Empty Recycle Bin" in the right-click menu.

If this item does not have an artificial shortcut key specified using &, then NVDA will automatically guess that its shortcut key is the first letter "E" of the first word.
However, for the Chinese interface, NVDA will regard the first Chinese character "清" of the five "清空回收站" as a shortcut.
This is weird and doesn't help.

Thanks

**Comment by cary-rowen** [comment]
@CyrilleB79 
Thanks for your answer, it's clear now.

**Review by michaelDCurran — CHANGES_REQUESTED** [review]
In general this approach looks okay to me.
However,  I found _splitShortcut a little confusing. If '+' (with or without whitespace is the only separator, You could pretty much get rid of the function, just split on '+', process each key with getKeySpeech (stripping any whitespace) and join it again with ' + '. 
However, as for _splitSequentialShortcut: I found this very hard to follow. At very least could you provide a dostring explaining exactly what it does? Then perhaps I may have morefeedback on how that one could be improved.

File: source/speech/speech.py [hunk 1]
```diff
@@ -1861,6 +1860,105 @@ def getPropertiesSpeech(  # noqa: C901
 	return textList
 
 
+def speakKeyboardShortcuts(keyboardShortcutsStr: Optional[str]) -> SpeechSequence:
+	speak(getKeyboardShortcutsSpeech(keyboardShortcutsStr))
```
Comment (seanbudd) [inline source/speech/speech.py:1864]: should this return here based on the return type?

```suggestion
	return speak(getKeyboardShortcutsSpeech(keyboardShortcutsStr))
```

Comment (CyrilleB79) [reply source/speech/speech.py:1864]: No, the return type was wrong. I have fixed it in 449e876.

File: source/speech/speech.py [hunk 2]
```diff
@@ -1861,6 +1860,127 @@ def getPropertiesSpeech(  # noqa: C901
...
+	for item in seq:
+		if len(seqOut) > 0 and isinstance(seqOut[-1], str) and isinstance(item, str):
+			seqOut[-1] = seqOut[-1] + item
+		else:
+			seqOut.append(item)
+
+	return seqOut
+
+
+def _getKeyboardShortcutSpeech(keyboardShortcut: str) -> SpeechSequence:
```
Comment (seanbudd) [inline source/speech/speech.py:1899]: it would be great if simple unit tests were added to these helper functions, just to document what they do.

File: source/speech/__init__.py [hunk 3]
```diff
@@ -46,6 +46,7 @@
 	RE_INDENTATION_SPLIT,
 	setSpeechMode,
 	speak,
+	speakKeyboardShortcuts,
```
[the remark this answers is not in the document]
Comment (CyrilleB79) [reply source/speech/__init__.py:49]: In `globalCommands.py`, `speech` is imported, not `speech.speech`. So I have thought that `speech` should be used out of the package, not `speech.speech`.
If it is not the case, I do not fully understand why `globalCommands.py` uses the variables in `speech` and not in `speech.speech`/

Anyway, if needed, I can probably use `speech.speech..speakKeyboardShortcuts` instead of `speech.speakKeyboardShortcuts` in `globalCommands.py`.

Let me know.

File: source/speech/speech.py [hunk 4]
```diff
@@ -1861,6 +1860,127 @@ def getPropertiesSpeech(  # noqa: C901
...
+	elif '+' in shortcut:
+		separator = '+'
+	else:
+		raise RuntimeError(f'The shortcut "{shortcut}" needs to contain a "+" symbol.')
+	if shortcut[-1] == '+':  # E.g. "Ctrl+Shift++"
+		keyList = shortcut[:-1].split(separator)
+		keyList[-1] = keyList[-1] + '+'
+	else:
+		keyList = shortcut.split(separator)
+	return keyList, separator
```
[the remark this answers is not in the document]
Comment (CyrilleB79) [reply source/speech/speech.py:1958]: Generally, I prefer keeping the string as defined by the provider and only add additional `CharacterMode` commands to indicate how characters should be pronounced.

In the case of the Magnifier's shortcut key indication ("Touche de logo Windows + Plus (+)"), if you copy the text (e.g. with speech history add-on), it's interesting to be able to differentiate the + key (appearing between parenthèses) and the + surrounded by spaces used to indicate a key combination. Maybe that's the reason why they used spaces around the + and IMO, it's interesting to keep this information, e.g. in braille.

File: source/speech/speech.py [hunk 5]
```diff
@@ -1861,6 +1860,127 @@ def getPropertiesSpeech(  # noqa: C901
...
+
+
+def _splitSequentialShortcut(shortcut: str) -> Tuple[List[str], List[str]]:
```
[the remark this answers is not in the document]
Comment (CyrilleB79) [reply source/speech/speech.py:1961]: Not at all. The comma and the space do not have the same meaning in Word's sequential shortcuts. The comma separates the tooltips seen visually in the ribbon. On each control of the ribbon, the tooltip is displayed with one or two letters. When there are two letters, they are separated by a space without comma.

Let's take an example:
In my French Word, in the "Accueil" (Home) ribbon, there is a button called "Effacer la mise en forme" (clear formatting). 
1. When I press Alt, all the tooltips in the upper ribbon appear visually; a "L" appears on "Accueil"
2. If I arrow left/right to "Accueil" ribbon, I can hear "Alt, L" when it is selected. That means that I can press "Alt, L" to get directly to it next time when the ribbon is not yet selected.
3. I select this ribbon
4. In the lower ribbon the tooltips are now displayed for all the controls. Among which "MF" for the button "Effacer la mise en forme". That means that if I press M then F at this point, the button will be activated. Or said otherwise, if I want to activate the button when the ribbon is not selected, I have to press Alt, then L then M then F.

So pressing "Alt, F" selects something and pressing "Alt, L, M F" selects something. But "Alt, L, M" does not do anything more than "Alt, L" since another character is expected. Hence the space between the M and the F.

File: source/speech/__init__.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (seanbudd) [reply source/speech/__init__.py:49]: >If it is not the case, I do not fully understand why globalCommands.py uses the variables in speech and not in speech.speech

I think this is because we didn't refactor usages of speech module when we performed the speech refactor. The goals was to keep it mostly backwards compatible, so internal and external code didn't need to be changed as much.

However, for simplicity, I think it's fine to keep in here.

**Review by seanbudd — APPROVED** [review]
Thanks @CyrilleB79, this LGTM

**Comment by seanbudd** [comment]
>Note: all the processing to split shortcut keys is in speech/speech.py as private functions. Reviewer: should they remain here or would they fit better elsewhere?

Hi sorry, just saw this note. I agree it's probably worth moving all the added functions to their own submodule, e.g. `speech.shortcutKeys`

**Comment by CyrilleB79** [comment]
> > Note: all the processing to split shortcut keys is in speech/speech.py as private functions. Reviewer: should they remain here or would they fit better elsewhere?
> 
> Hi sorry, just saw this note. I agree it's probably worth moving all the added functions to their own submodule, e.g. `speech.shortcutKeys`

Public functions `speakKeyboardShortcuts` and `getKeyboardShortcutsSpeech` are made with the same naming convention than other ones found in `speech.speech.py`, e.g. `speakObject` and `getObjectSpeech`. Would it make sense to keep them in `speech.speech.py` or should I also move them in the new file `speech.shortcutKeys.py`?
Should I move only private functions (prefixed with "_") instead?

Moreover, in `_getKeySpeech`, I call `speech.speech.getCurrentLanguage` which is defined in `speech.speech.py`. Moving this function as is in a separate file may cause cyclic import issues.

An alternative may be to pass the language as parameter, but in this case, I need to pass it as parameter in `_getKeySpeech` and `_getKeyboardShortcutSpeech`. This seems a bit ugly.

What do you think?

If you expect also the public functions to be moved in `speech.shortcutKeys.py`, more functions would need to be passed the language as parameter, which seems worse.

**Comment by seanbudd** [comment]
>Public functions speakKeyboardShortcuts and getKeyboardShortcutsSpeech are made with the same naming convention than other ones found in speech.speech.py, e.g. speakObject and getObjectSpeech. Would it make sense to keep them in speech.speech.py or should I also move them in the new file speech.shortcutKeys.py?
Should I move only private functions (prefixed with "_") instead?

Thanks, I think moving all new functions would be ideal. The goal is to move functions out of `speech.speech` rather than add to it. e.g. a new submodule of `speech.NVDAObjects` would be ideal.

>Moreover, in _getKeySpeech, I call speech.speech.getCurrentLanguage which is defined in speech.speech.py. Moving this function as is in a separate file may cause cyclic import issues.

You should be able to lazy import this in `_getKeySpeech`

**Comment by CyrilleB79** [comment]
> Thanks, I think moving all new functions would be ideal. The goal is to move functions out of `speech.speech` rather than add to it. e.g. a new submodule of `speech.NVDAObjects` would be ideal.

Done in 5ff17f7890e9bcab8635a9ae0dae99eaf9a0e379. While at it, I have also added one more unit test on `getKeyboardShortcutsSpeech` to capture the case when the shortcut key string contains two shortcuts.

> > Moreover, in _getKeySpeech, I call speech.speech.getCurrentLanguage which is defined in speech.speech.py. Moving this function as is in a separate file may cause cyclic import issues.
> 
> You should be able to lazy import this in `_getKeySpeech`

Yes. I have done this needing two lazy imports. IMO, using lazy imports show a suboptimal code organization. But since splitting `speech.py` would be out-of-scope of this PR (and I do not intend to do it), let's keep them.

**Review by seanbudd — APPROVED** [review]
Thanks @CyrilleB79

**Comment by michaelDCurran** [comment]
This PR has been reverted in #15237 

Unfortunately this PR causes Code Factory Eloquence to spell out menu items with shortcut keys, such as in the NVDA menu.
Although the bug is technically in the Code Factory driver I think - not handlin character mode command correctly when embedded in other speech,
this PR is the first known case that exercises the bug directly.
Because 2023.2 is not an add-on breaking release, we are reverting pr #14900.
A replacement pr for 2023.3 can be opened that avoids affecting Code Factory Eloquence if at all possible, or a new copy of the reverted pr can be opened for NVDA 2024.1.

**Comment by CyrilleB79** [comment]
In anticipation of the reintroduction of this PR in NVDA 2024.1, the following code snippet executed in the console should allow Code Factory developers to test if their driver correctly supports `CharacterModeCommand`.

```
import core
from speech import speak
from speech.commands import CharacterModeCommand

# The following lines executed in the console may not pronounce the sentence correctly due to
# a possible confusion between "a" spoken as a character and "a" spoken as a word.
seq1 = ['The letter ', 'a', 'spoken as a character should not be pronounced as the word a spoken as a whole word.']
core.callLater(1000, lambda: speech.speak(seq1)) and None

# The following lines executed in the console should pronounce the sentence correctly due to
# explicit indication of when "a" should be poken as a character (letter).
seq2 = ['The letter ', CharacterModeCommand(True), 'a', CharacterModeCommand(False), 'spoken as a character should not be pronounced as the word a spoken as a whole word.']
core.callLater(1000, lambda: speech.speak(seq2)) and None
```

@michaelDCurran have you had a contact with CodeFactory regarding this issue? Do they confirm that the bug is in their current driver?

**Comment by CyrilleB79** [comment]
@seanbudd, @michaelDCurran
Now that we have entered the 2024.1 dev cycle, may I reintroduce the changes of this PR? Should I do it now or should I wait later (and when) in the dev cycle?
If I introduce it now, it will break the UX for alpha users using Code Factory synthesizer; but it will provide Code Factory devs early a version on which they can work to fix the issue.
Thanks.

**Comment by seanbudd** [comment]
Yep, opening a PR for this would be welcome.
